### PR TITLE
Add basic cmake support

### PR DESCRIPTION
**Pull request by ScaledLizard**
Please check if you can accept my pull request. It adds basic cmake compilation support for libbfio. I have tested it with Microsoft Visual Studio, but the most import other compilers (gcc, clang) should work as well.

### CODE REVIEW

**Comment by joachimmetz** [comment]
Per https://github.com/libyal/libfmapi/issues/4 "I'd need a solution that works for the whole of libyal and that can reliably replace autotools." this should include:

* generating source from .l and .y files
* generating files from template files (.in)
* building binaries, including nearly static builds
* building Python bindings
* building OSSFuzz fuzz targets
* integration with gettext or other i18n/l10n support
* CI tests
* support for specific make targets like `make distclean`, `make sources`

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+../libbfio/libbfio_file_range.c
+../libbfio/libbfio_file_range_io_handle.c
+../libbfio/libbfio_handle.c
+../libbfio/libbfio_memory_range.c
+../libbfio/libbfio_memory_range_io_handle.c
+../libbfio/libbfio_pool.c
+../libbfio/libbfio_support.c
+../libbfio/libbfio_system_string.c)
+
+target_include_directories(libbfio PRIVATE
```
Comment (joachimmetz) [inline cmake/CMakeLists.txt:18]: This is not portable, this expects dependencies to be installed in the same directory as the libbfio source

**Comment by joachimmetz** [comment]
Closing this for now, until a more portable and comprehensive solution is provided.

**Comment by ScaledLizard** [comment]
That is a long list of requirements. It may be possible to meet some or most of them, but if a single one will cause the project to fail, that is very discouraging. I would still plead that providing cmake support is more flexible than the Visual Studio files that are currently provided. For people who know cmake better than autoconf/automake (like me), projects with cmake support can easily be configured to be compiled with several compilers using a single configuration, and the output files of several compilers can coexist easily. I would even argue that cmake support would be a good thing for the project even if it's not continually maintained - in which case, the file could reside in a folder named "unmaintained".

About the individual things you mentioned:

- "generating source from .l and .y files"
  cmake supports running external tools via execute_process
- "generating files from template files (.in)"
  I would put the required definitions in CMakeLists.txt. If necessary, it is possible to set operating system- and compiler-specific definitions there.
- "building binaries, including nearly static builds"
  I suppose you are discussing whether to link dependencies statically or dynamically. Yes, cmake supports that.
- "building Python bindings"
  I suppose you want to integrate libyal into Python at runtime (not the other way around). I would imagine that you would produce a .dll or .so to load into Python, and I suppose that will work - I have made a Windows .dll to load as a plugin to run Python scripts inside our application. I suppose all you need for this is a .dll or .so, so I guess it will work.
- "building OSSFuzz fuzz targets"
  I'm not sure exactly this means, sorry.
- integration with gettext or other i18n/l10n support
  gettext is a GNU library that can be linked into executables using cmake, sure.
- CI tests
  This requires building a test apllication (possible), running it via execute_process (possible), and then setting a variable to contain the result:
  https://stackoverflow.com/questions/52076948/check-exit-status-of-execute-process-in-cmake
  I didn't test this myself yet, but others report that it works.
- "support for specific make targets like make distclean, make sources"
  I usually run cmake to create build files in a separate directory, which I can completely delete when I'm done or if I want to start over from a clean state.
  However, it is possible to add custom targets to cmake using the command add_custom_target.
- "this expects dependencies to be installed in the same directory as the libbfio source"
  It would be possible to have variables in CMakeLists.txt that point to the include directories of the individual packages. There could even be switches to select a default location for these folders, like /usr/include, if needed. Such user-specific settings can even be in an external file that is included via the "include" command.

However, if you don't want this, I'll drop the issue.

**Comment by joachimmetz** [comment]
> I suppose you are discussing whether to link dependencies statically or dynamically. Yes, cmake supports that.

I mean helping with the build parameters

**Comment by joachimmetz** [comment]
> However, if you don't want this, I'll drop the issue.

Having 2 build systems is costly, one other option is add cmake support to https://github.com/libyal/vstools

**Comment by joachimmetz** [comment]
> I'm not sure exactly this means, sorry.

E.g. https://github.com/libyal/libpff/tree/main/ossfuzz

**Pull request closed without merging**